File: tools-ts/src/splitter.ts
/**
 * Split an esbuild bundle into individual module files.
 *
 * Uses the TypeScript parser to properly handle string literals,
 * template literals, regex, and nested scopes — no fragile regex.
 *
 * Detects CJS (d) and ESM (G) module wrapper patterns from the preamble,
 * then extracts each `var NAME = WRAPPER((...) => {...})` declaration.
 */

import ts from "typescript";
import * as fs from "fs";
import * as path from "path";

export interface Section {
  index: number;
  filename: string;
  type: "preamble" | "section" | "tail";
  size: number;
  start: number;
  end: number;
  moduleName?: string;
  moduleKind?: string;
  hasGlue?: boolean;
  glueSize?: number;
}

export interface Manifest {
  sourceFile: string;
  sourceSize: number;
  sectionCount: number;
  wrapperNames: { cjs: string | null; esm: string | null };
  sections: Section[];
}

/**
 * Auto-detect CJS and ESM wrapper function names from the preamble.
 *
 * CJS pattern: var d=(H,_)=>()=>(_||H((_={exports:{}}).exports,_),_.exports)
 * ESM pattern: var G=(H,_)=>()=>(H&&(_=H(H=0)),_)
 */
function detectWrapperNames(source: string): { cjs: string | null; esm: string | null } {
  const preamble = source.slice(0, 5000);

  // CJS: contains {exports:{}} pattern — may start with var or ,
  const cjsMatch = preamble.match(
    /(?:var|,)\s*(\w{1,3})\s*=\s*\(\w,\w\)\s*=>\s*\(\)\s*=>\s*\(\w\|\|\w\(\(\w=\{exports:\{/
  );

  // ESM: the &&/=0 pattern — may start with var or ,
  const esmMatch = preamble.match(
    /(?:var|,)\s*(\w{1,3})\s*=\s*\(\w,\w\)\s*=>\s*\(\)\s*=>\s*\(\w&&\(\w=\w\(\w=0\)\)/
  );

  return {
    cjs: cjsMatch?.[1] ?? null,
    esm: esmMatch?.[1] ?? null,
  };
}

/**
 * Find the matching closing paren/brace using the TS scanner.
 * More reliable than manual parsing since it handles all JS syntax.
 */
function findMatchingParen(source: string, openPos: number): number {
  let depth = 0;
  const openChar = source[openPos];
  const closeChar = openChar === "(" ? ")" : openChar === "{" ? "}" : "]";

  // Use a simple stack-based approach with string awareness
  let i = openPos;
  let inString: string | null = null;

  while (i < source.length) {
    const c = source[i];

    if (inString) {
      if (c === "\\" ) { i += 2; continue; }
      if (c === inString) inString = null;
      i++;
      continue;
    }

    if (c === '"' || c === "'" || c === "`") {
      inString = c;
      i++;
      continue;
    }

    // Skip comments
    if (c === "/" && i + 1 < source.length) {
      if (source[i + 1] === "/") {
        const nl = source.indexOf("\n", i);
        i = nl === -1 ? source.length : nl + 1;
        continue;
      }
      if (source[i + 1] === "*") {
        const end = source.indexOf("*/", i + 2);
        i = end === -1 ? source.length : end + 2;
        continue;
      }
    }

    if (c === openChar) depth++;
    else if (c === closeChar) {
      depth--;
      if (depth === 0) return i;
    }

    i++;
  }

  return -1;
}

/**
 * Extract string hints from module body for the manifest.
 */
function extractHints(body: string): {
  requires?: string[];
  strings?: string[];
  errors?: string[];
} {
  const hints: Record<string, string[]> = {};

  // require() calls
  const requires = [...body.matchAll(/require\("([^"]+)"\)/g)].map((m) => m[1]);
  if (requires.length) hints.requires = [...new Set(requires)].slice(0, 10);

  // Significant strings
  const strings = [...body.matchAll(/"([A-Za-z][\w\-\./ ]{5,80})"/g)]
    .map((m) => m[1])
    .filter((s) => !["object", "function", "string", "number", "create"].includes(s));
  const uniqueStrings = [...new Set(strings)].slice(0, 10);
  if (uniqueStrings.length) hints.strings = uniqueStrings;

  // Error messages
  const errors = [
    ...body.matchAll(/(?:Error|throw|error)\s*\(\s*["`]([^"`]{10,100})/g),
  ].map((m) => m[1]);
  if (errors.length) hints.errors = [...new Set(errors)].slice(0, 5);

  return hints;
}

export function splitSource(sourcePath: string, outputDir: string): Manifest {
  const source = fs.readFileSync(sourcePath, "utf-8");
  console.log(`Reading ${sourcePath}...`);
  console.log(`  ${source.length.toLocaleString()} chars`);

  const wrappers = detectWrapperNames(source);

  // Fallback: detect by frequency analysis if auto-detection fails
  if (!wrappers.cjs || !wrappers.esm) {
    // Only consider names that are defined as thunk factories: var X=(a,b)=>()=>
    const thunkDefs = new Set<string>();
    for (const m of source.slice(0, 5000).matchAll(/var (\w{1,3})=\(\w,\w\)=>\(\)=>/g)) {
      thunkDefs.add(m[1]);
    }
    // Count how often each thunk is used as a module wrapper: var NAME=WRAPPER((
    const freq = new Map<string, number>();
    for (const m of source.matchAll(/var \w+=(\w{1,3})\(\(/g)) {
      if (thunkDefs.has(m[1])) freq.set(m[1], (freq.get(m[1]) || 0) + 1);
    }
    const sorted = [...freq.entries()].sort((a, b) => b[1] - a[1]);
    if (sorted.length >= 1 && !wrappers.esm) wrappers.esm = sorted[0][0];
    if (sorted.length >= 2 && !wrappers.cjs) wrappers.cjs = sorted[1][0];
  }

  console.log(`  Wrapper functions: CJS=${wrappers.cjs}, ESM=${wrappers.esm}`);

  const wrapperNames = [wrappers.cjs, wrappers.esm].filter(Boolean);
  if (!wrapperNames.length) {
    throw new Error("No wrapper functions detected");
  }

  // Find all module declarations: var NAME = WRAPPER((
  const wrapperPattern = new RegExp(
    `var (\\w+)=(${wrapperNames.map((n) => n!.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")).join("|")})\\(\\(`,
    "g"
  );

  const modules: Array<{
    name: string;
    kind: string;
    start: number;
    end: number;
  }> = [];

  let match: RegExpExecArray | null;
  let lastModuleEnd = 0;
  while ((match = wrapperPattern.exec(source)) !== null) {
    // Skip if this match starts inside a previous module (overlap)
    if (match.index < lastModuleEnd) continue;

    // Pattern matched: var NAME=WRAPPER((
    // match[0] ends at the second (, so the first ( is at length-2, second at length-1
    const outerParen = match.index + match[0].length - 2; // WRAPPER( ← this one
    const closeParen = findMatchingParen(source, outerParen);
    if (closeParen === -1) continue;

    const moduleSize = closeParen - outerParen;

    // Validation: inner content should start with a function factory pattern
    // Real: WRAPPER((params) => {...})  or WRAPPER(() => {...})
    // False: WRAPPER((someValue))
    const innerParen = outerParen + 1; // The second ( — start of the factory function params
    const innerSample = source.slice(innerParen, Math.min(innerParen + 200, closeParen));
    const looksLikeModule = /^\([^)]*\)\s*=>/.test(innerSample) ||
                            /^\(\s*function/.test(innerSample) ||
                            innerSample.includes("function");
    if (!looksLikeModule) continue;

    // Sanity: skip impossibly large modules (>500KB)
    if (moduleSize > 500_000) continue;

    let moduleEnd = closeParen + 1;
    if (moduleEnd < source.length && source[moduleEnd] === ";") moduleEnd++;

    // Must not overlap with previous module
    if (match.index < lastModuleEnd) continue;

    modules.push({
      name: match[1],
      kind: match[2],
      start: match.index,
      end: moduleEnd,
    });
    lastModuleEnd = moduleEnd;
  }

  console.log(`  ${modules.length} modules found`);

  // Build sections: preamble (before first module), then glue+module pairs, then tail
  const sections: Section[] = [];
  let pos = 0;
  let idx = 0;

  // Preamble: everything before the first module
  if (modules.length > 0 && modules[0].start > 0) {
    sections.push({
      index: idx,
      filename: `${String(idx).padStart(4, "0")}_preamble.js`,
      type: "preamble",
      size: modules[0].start,
      start: 0,
      end: modules[0].start,
    });
    idx++;
    pos = modules[0].start;
  }

  for (const mod of modules) {
    // Include any glue code between previous module end and this module start
    const hasGlue = mod.start > pos;
    const sectionStart = hasGlue ? pos : mod.start;

    sections.push({
      index: idx,
      filename: `${String(idx).padStart(4, "0")}_${mod.name}.js`,
      type: "section",
      size: mod.end - sectionStart,
      start: sectionStart,
      end: mod.end,
      moduleName: mod.name,
      moduleKind: mod.kind,
      hasGlue,
      glueSize: hasGlue ? mod.start - pos : undefined,
    });
    idx++;
    pos = mod.end;
  }

  // Tail
  if (pos < source.length) {
    sections.push({
      index: idx,
      filename: `${String(idx).padStart(4, "0")}_tail.js`,
      type: "tail",
      size: source.length - pos,
      start: pos,
      end: source.length,
    });
  }

  // Verify coverage
  const totalSize = sections.reduce((sum, s) => sum + s.size, 0);
  if (totalSize !== source.length) {
    throw new Error(`Coverage mismatch: ${totalSize} vs ${source.length}`);
  }

  // Write files
  fs.mkdirSync(outputDir, { recursive: true });

  for (const section of sections) {
    const body = source.slice(section.start, section.end);
    fs.writeFileSync(path.join(outputDir, section.filename), body);

    // Add hints for sections
    if (section.type === "section") {
      const hints = extractHints(body);
      (section as any).hints = hints;
    }
  }

  // Write manifest
  const manifest: Manifest = {
    sourceFile: path.basename(sourcePath),
    sourceSize: source.length,
    sectionCount: sections.length,
    wrapperNames: wrappers,
    sections,
  };

  fs.writeFileSync(
    path.join(outputDir, "_manifest.json"),
    JSON.stringify(manifest, null, 2)
  );

  console.log(`\nWrote ${sections.length} sections to ${outputDir}/`);
  console.log(`  Preamble: ${sections.filter((s) => s.type === "preamble").length}`);
  console.log(`  Modules: ${sections.filter((s) => s.type === "section").length}`);
  console.log(`  Tail: ${sections.filter((s) => s.type === "tail").length}`);

  return manifest;
}

// CLI entry
if (import.meta.main) {
  const args = process.argv.slice(2);
  if (args.length < 1) {
    console.log("Usage: bun run src/splitter.ts <source.js> [output_dir]");
    process.exit(1);
  }
  splitSource(args[0], args[1] || "modules");
}

File: tools/splitter.py
#!/usr/bin/env python3
"""
Split a bundled Claude Code JS source into individual module files.

Parses the esbuild bundle format, identifying:
- Runtime preamble (helper functions)
- CJS modules: CommonJS-style lazy module factories (auto-detected)
- ESM modules: ESM-style lazy initializers (auto-detected)
- Glue code: top-level declarations and functions between modules
- Tail: entry point code after the last module

Wrapper function names are detected automatically from the preamble,
so this doesn't need updating when esbuild minification changes them.

Outputs numbered section files + a manifest.json for reassembly.
"""

import json
import os
import re
import sys


def find_matching_paren(data, open_pos):
    """Find the ) that matches the ( at open_pos.

    Properly handles:
    - Single/double quoted strings with escapes
    - Template literals with nested ${expressions}
    - Regular expressions (with character classes)
    - Line and block comments
    """
    paren_depth = 0
    i = open_pos

    # Context stack: 'sq', 'dq', 'tpl', 'tpl_expr', 'brace'
    stack = []

    while i < len(data):
        c = data[i]
        top = stack[-1] if stack else None

        # --- Inside single-quoted string ---
        if top == "sq":
            if c == "\\":
                i += 2
                continue
            if c == "'":
                stack.pop()
            i += 1
            continue

        # --- Inside double-quoted string ---
        if top == "dq":
            if c == "\\":
                i += 2
                continue
            if c == '"':
                stack.pop()
            i += 1
            continue

        # --- Inside template literal ---
        if top == "tpl":
            if c == "\\":
                i += 2
                continue
            if c == "`":
                stack.pop()
                i += 1
                continue
            if c == "$" and i + 1 < len(data) and data[i + 1] == "{":
                stack.append("tpl_expr")
                i += 2
                continue
            i += 1
            continue

        # --- Inside template expression ${...} or nested brace ---
        if top in ("tpl_expr", "brace"):
            if c == "'":
                stack.append("sq")
                i += 1
                continue
            if c == '"':
                stack.append("dq")
                i += 1
                continue
            if c == "`":
                stack.append("tpl")
                i += 1
                continue
            if c == "/" and i + 1 < len(data):
                if data[i + 1] == "/":
                    nl = data.find("\n", i)
                    i = nl + 1 if nl != -1 else len(data)
                    continue
                if data[i + 1] == "*":
                    end = data.find("*/", i + 2)
                    i = end + 2 if end != -1 else len(data)
                    continue
                if _is_regex_start(data, i):
                    i = _skip_regex(data, i)
                    continue
            if c == "{":
                stack.append("brace")
                i += 1
                continue
            if c == "}":
                stack.pop()
                i += 1
                continue
            i += 1
            continue

        # --- Top-level parsing ---
        if c == "'":
            stack.append("sq")
            i += 1
            continue
        if c == '"':
            stack.append("dq")
            i += 1
            continue
        if c == "`":
            stack.append("tpl")
            i += 1
            continue

        # Comments
        if c == "/" and i + 1 < len(data):
            if data[i + 1] == "/":
                nl = data.find("\n", i)
                i = nl + 1 if nl != -1 else len(data)
                continue
            if data[i + 1] == "*":
                end = data.find("*/", i + 2)
                i = end + 2 if end != -1 else len(data)
                continue
            if _is_regex_start(data, i):
                i = _skip_regex(data, i)
                continue

        if c == "(":
            paren_depth += 1
        elif c == ")":
            paren_depth -= 1
            if paren_depth == 0:
                return i

        i += 1
    return -1


def _is_regex_start(data, pos):
    """Check if / at pos starts a regex literal (not division)."""
    j = pos - 1
    while j >= 0 and data[j] in " \t\n\r":
        j -= 1
    if j < 0:
        return True
    c = data[j]
    if c in "(,=:;[!&|?{+->~^%<":
        return True
    for kw in (
        "return", "typeof", "void", "delete", "throw",
        "new", "case", "instanceof", "in",
    ):
        end = j + 1
        start = end - len(kw)
        if start >= 0 and data[start:end] == kw:
            before = start - 1
            if before < 0 or not (data[before].isalnum() or data[before] in "_$"):
                return True
    return False


def _skip_regex(data, pos):
    """Skip past a regex literal /.../ including flags."""
    i = pos + 1
    in_class = False
    while i < len(data):
        c = data[i]
        if c == "\\":
            i += 2
            continue
        if in_class:
            if c == "]":
                in_class = False
            i += 1
            continue
        if c == "[":
            in_class = True
            i += 1
            continue
        if c == "/":
            i += 1
            while i < len(data) and data[i].isalpha():
                i += 1
            return i
        i += 1
    return i


def extract_hints(data, body_text, max_strings=10):
    """Extract hints about a module's purpose from its content."""
    hints = {}

    # require() calls
    requires = re.findall(r'require\("([^"]+)"\)', body_text[:5000])
    if requires:
        hints["requires"] = list(dict.fromkeys(requires))[:10]

    # Significant string literals (skip very short/generic ones)
    strings = re.findall(r'"([A-Za-z][\w\-\./ ]{5,80})"', body_text[:10000])
    # Deduplicate, keep order
    seen = set()
    unique = []
    for s in strings:
        if s not in seen and s not in ("object", "function", "string", "number", "create"):
            seen.add(s)
            unique.append(s)
    if unique:
        hints["strings"] = unique[:max_strings]

    # Error messages
    errors = re.findall(r'(?:Error|throw|error)\s*\(\s*["`]([^"`]{10,100})', body_text[:10000])
    if errors:
        hints["errors"] = list(dict.fromkeys(errors))[:5]

    return hints


def detect_wrapper_names(data, preamble_size=5000):
    """Auto-detect the CJS and ESM module wrapper function names from the preamble.

    CJS pattern: varname=(a,b)=>()=>(b||a((b={exports:{}}).exports,b),b.exports)
    ESM pattern: varname=(a,b)=>()=>(a&&(b=a(a=0)),b)
    """
    preamble = data[:preamble_size]

    V = r'[\w$]{1,3}'  # JS variable name (1-3 chars, includes $ and _)

    # CJS: contains {exports:{}} - the factory that creates module objects
    cjs_match = re.search(rf'({V})=\({V},{V}\)=>\(\)=>\({V}\|\|{V}\(\({V}=\{{exports:\{{', preamble)
    # ESM: the simpler lazy initializer with the &&/=0 pattern
    esm_match = re.search(rf'({V})=\({V},{V}\)=>\(\)=>\({V}&&\({V}={V}\({V}=0\)\)', preamble)

    cjs_name = cjs_match.group(1) if cjs_match else None
    esm_name = esm_match.group(1) if esm_match else None

    if not cjs_name or not esm_name:
        print(f"  WARNING: Could not auto-detect wrapper names (CJS={cjs_name}, ESM={esm_name})", file=sys.stderr)
        print(f"  Falling back to searching for most common var X=(( patterns...", file=sys.stderr)
        # Fallback: find the most common short identifiers used in var X=Y(( patterns
        from collections import Counter
        candidates = re.findall(r'var \w+=(\w{1,3})\(\(', data[:50000])
        counts = Counter(candidates).most_common(2)
        if len(counts) >= 2:
            # The more common one is ESM (more modules), less common is CJS
            esm_name = esm_name or counts[0][0]
            cjs_name = cjs_name or counts[1][0]
        elif len(counts) == 1:
            esm_name = esm_name or counts[0][0]

    print(f"  Wrapper functions: CJS={cjs_name}, ESM={esm_name}")
    return cjs_name, esm_name


def split_source(source_path, output_dir):
    """Split source.js into individual section files."""
    print(f"Reading {source_path}...")
    data = open(source_path, "r").read()
    print(f"  {len(data):,} chars")

    # Auto-detect module wrapper function names from the preamble
    cjs_name, esm_name = detect_wrapper_names(data)
    wrapper_names = "|".join(filter(None, [re.escape(cjs_name) if cjs_name else None,
                                            re.escape(esm_name) if esm_name else None]))
    if not wrapper_names:
        print("  ERROR: No wrapper functions detected, cannot split.", file=sys.stderr)
        sys.exit(1)

    module_pattern = re.compile(rf"var (\w+)=({wrapper_names})\(\(")
    matches = list(module_pattern.finditer(data))
    print(f"  {len(matches)} modules found")

    # Build segment list
    segments = []
    pos = 0

    for m in matches:
        open_paren = m.end() - 2  # outer ( of F(( or G((
        close_paren = find_matching_paren(data, open_paren)
        if close_paren == -1:
            print(f"  WARNING: Failed to parse module {m.group(1)} at {m.start()}", file=sys.stderr)
            continue

        mod_end = close_paren + 1
        if mod_end < len(data) and data[mod_end] == ";":
            mod_end += 1

        if m.start() > pos:
            segments.append({"type": "glue", "start": pos, "end": m.start()})

        segments.append({
            "type": "module",
            "name": m.group(1),
            "kind": m.group(2),
            "start": m.start(),
            "end": mod_end,
        })
        pos = mod_end

    if pos < len(data):
        segments.append({"type": "tail", "start": pos, "end": len(data)})

    # Group into sections: [glue] + [module]
    # First glue is always preamble; subsequent glues merge with their module
    sections = []
    i = 0
    while i < len(segments):
        seg = segments[i]
        if seg["type"] == "glue":
            if i == 0 and seg["start"] == 0:
                # First glue is always preamble
                sections.append({
                    "type": "preamble",
                    "start": seg["start"],
                    "end": seg["end"],
                })
                i += 1
            elif i + 1 < len(segments) and segments[i + 1]["type"] == "module":
                mod = segments[i + 1]
                sections.append({
                    "type": "section",
                    "glue_start": seg["start"],
                    "glue_end": seg["end"],
                    "module_name": mod["name"],
                    "module_kind": mod["kind"],
                    "start": seg["start"],
                    "end": mod["end"],
                })
                i += 2
            else:
                # Standalone glue (orphan)
                sections.append({
                    "type": "preamble",
                    "start": seg["start"],
                    "end": seg["end"],
                })
                i += 1
        elif seg["type"] == "module":
            # Module with no preceding glue
            sections.append({
                "type": "section",
                "glue_start": None,
                "glue_end": None,
                "module_name": seg["name"],
                "module_kind": seg["kind"],
                "start": seg["start"],
                "end": seg["end"],
            })
            i += 1
        elif seg["type"] == "tail":
            sections.append({
                "type": "tail",
                "start": seg["start"],
                "end": seg["end"],
            })
            i += 1
        else:
            i += 1

    # Verify complete coverage
    total = sum(s["end"] - s["start"] for s in sections)
    assert total == len(data), f"Coverage mismatch: {total} vs {len(data)}"

    # Verify contiguity
    for j in range(len(sections) - 1):
        assert sections[j]["end"] == sections[j + 1]["start"], \
            f"Gap between sections {j} and {j+1}: {sections[j]['end']} != {sections[j+1]['start']}"

    # Write section files
    os.makedirs(output_dir, exist_ok=True)

    manifest = {
        "source_file": os.path.basename(source_path),
        "source_size": len(data),
        "section_count": len(sections),
        "sections": [],
    }

    for idx, section in enumerate(sections):
        body = data[section["start"]:section["end"]]
        sec_type = section["type"]

        if sec_type == "preamble":
            filename = f"{idx:04d}_preamble.js"
        elif sec_type == "tail":
            filename = f"{idx:04d}_tail.js"
        else:
            mod_name = section.get("module_name", "unknown")
            mod_kind = section.get("module_kind", "?")
            filename = f"{idx:04d}_{mod_name}.js"

        filepath = os.path.join(output_dir, filename)
        open(filepath, "w").write(body)

        entry = {
            "index": idx,
            "filename": filename,
            "type": sec_type,
            "size": len(body),
        }

        if sec_type == "section":
            entry["module_name"] = section["module_name"]
            entry["module_kind"] = section["module_kind"]
            entry["has_glue"] = section["glue_start"] is not None
            if section["glue_start"] is not None:
                entry["glue_size"] = section["glue_end"] - section["glue_start"]

            # Extract hints
            hints = extract_hints(data, body)
            if hints:
                entry["hints"] = hints

        manifest["sections"].append(entry)

    # Write manifest
    manifest_path = os.path.join(output_dir, "_manifest.json")
    open(manifest_path, "w").write(json.dumps(manifest, indent=2))

    print(f"\nWrote {len(sections)} sections to {output_dir}/")
    print(f"  Preamble: {sum(1 for s in sections if s['type'] == 'preamble')}")
    print(f"  Modules: {sum(1 for s in sections if s['type'] == 'section')}")
    print(f"  Tail: {sum(1 for s in sections if s['type'] == 'tail')}")
    print(f"  Manifest: {manifest_path}")

    return manifest


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <source.js> [output_dir]")
        sys.exit(1)

    source = sys.argv[1]
    output = sys.argv[2] if len(sys.argv) > 2 else "modules"
    split_source(source, output)

